### DISCUSSION TITLE
Problems with an ipxe script

### DISCUSSION THREAD

**Discussion by DungeonMtr** [ipxe/ipxe#1230] [Q&A]
Hello everyone.

For the last days i've been working on a script in order to configure manually the pxe boot how we need it. So far it works just how we want it to work, however, we've encountered a somewhat inconsistent problem that seems to be linked with the rng. Sometimes, when booted with out ipxe script, the installation gets stuck in generating random bits, img for reference:

![image](https://github.com/ipxe/ipxe/assets/64873468/56b8b9f9-3d15-4c94-ad2c-926a13cf0b30)

Whenever this happens, the installation gets completely stuck and doesn't move on, and this seems to happen randomly and we haven't been able to determine the reason of why this happens. Here i'll leave the ipxe script.

```
#!ipxe

# Define variables
set index:int32 0
set open_index:int32 0

# Set the desired mac
set desired-mac:hex fa:16:3e:5b:3e:43

# Start the net interfaces
goto open_interfaces

# =====================FUNCTIONS=====================

# Tries to start all of the net interfaces and then goes to scan_interfaces.
:open_interfaces
  ifopen net${open_index} || goto scan_interfaces
  inc open_index
  goto open_interfaces

# Search the net interfaces for the desired mac.
:scan_interfaces
  isset ${net${index}/mac} && echo trying ${net${index}} || goto no_more_interfaces
  iseq ${net${index}/mac} ${desired-mac} && goto interface_found || echo ${net${index}} doesn't match the required MAC.
  inc index
  goto scan_interfaces

# Can't find the specified Mac.
:no_more_interfaces
  echo no aviable matches for MAC ${desired-mac}
  goto exit

# The specified Mac has been found and configures the Mac's settings with the desired configuration.
:interface_found
  echo Match found for MAC ${desired-mac} in net${index}
  set net${index}/ip 192.168.64.29
  set net${index}/netmask 255.255.255.0
  set net${index}/gateway 192.168.64.1
  goto close_interfaces
  goto open_desired_interface

# Attempts to close all net interfaces.
:close_interfaces
  echo closing net interfaces...
  ifclose || echo close_interfaces ERROR
  echo all aviable interfaces have been closed.

# Attempts to start the desired net interface
:open_desired_interface
  echo opening net${index}...
  ifopen net${index} || echo Open_desired_interface ERROR
  echo interfaz net${index} has been opened successfully.
  goto print_net

# Shows all the net configuration
:print_net
  echo net${index} has been configured...
  echo >ip: ${net${index}/ip}
  echo >mac: ${net${index}/mac}
  echo >netmask: ${net${index}/netmask}
  echo >gateway: ${net${index}/gateway}
  echo ===============================
  ifstat
  goto load_kernel_initrd

#Load the Kernel and the Initrd
:load_kernel_initrd
  kernel http://192.168.64.1/RHES86/images/pxeboot/vmlinuz inst.repo=http://192.168.64.1/RHES86/  inst.ks=http://192.168.64.1/pxe/kickstart/ksfiles/ks_idasipxe_192.168.64.29_RHES86.cfg && echo Succesfully loaded the kernel || echo ERROR in kernel
  initrd http://192.168.64.1/RHES86/images/pxeboot/initrd.img && echo Succesfully loaded the initrd || echo ERROR in initrd
  boot
```

Any help regarding this issue would be appreciated, Thanks!!

**Comment by NiKiZe** [comment]
The image shows a booted Linux instance (not iPXE) with RNG started as last successful.

After this 8021q and IPv6 kernel messages is shown.
To me this indicates that the interfaces might not be up yet.

Script mentions RHES86

We have no indication of network
You will probably have to debug the Linux boot, checking if interfaces gets IP etc.

I would also consider checking the switch port for STP configuration etc. And maybe even connections (bad cable causing link to take time to negotiate link speed when it fails at gigabit or similar)

**Reply by DungeonMtr** [reply]
First of all, thank you very much for your reply!
Secondly, i missed saying that the image with the script.ipxe is made with a dd on the booting disk. If you use the official ipxe.org image or you use the normal pxe process with the exact same image and ks, this never happens.
I understand that this is an OS problem. In this case, where it fails randomly are some openstack virtual machines. And the dmesg messages seen in the log always appear regardless of whether it works or not. Thats why i suspect that it might be a bad configuration in the ipxe script.

**Reply by mcb30** [reply]
There is nothing in what you've reported so far to suggest any kind of problem on the iPXE side.  It sounds as though you are seeing iPXE always reliably load and start the kernel, and that the problem occurs later while systemd is bringing the system up.

**Comment by ZeroQI** [comment]
I had some issue with loading some linux and freedos, where i needed [imgfree ](https://ipxe.org/cmd/imgfree) before loading the kernel, but I have a pictured background and scripts and multiple variables calling other scripts, might be worth try...
